### PR TITLE
Always use ExecuteReplicated with SPMD

### PR DESCRIPTION
**Pull request by jonb377**
SPMD computations should always follow the ExecuteReplicated codepath, even when there are no sharding annotations present in the HLO.

### CODE REVIEW

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 1]
```diff
@@ -1329,8 +1329,10 @@ XLAGraphExecutor::CompilationResult XLAGraphExecutor::Compile(
         torch::lazy::Output(ir_value.node.get(), ir_value.index));
     lowering_ctx.AddResult(root);
   }
+  // Always execute sharded when running in SPMD mode
+  bool is_sharded = (coll.device == GetVirtualDevice());
   // Annotate HLO sharding selectively in the compuation.
-  bool is_sharded = ShardingUtil::SetHloSharding(&lowering_ctx);
+  ShardingUtil::SetHloSharding(&lowering_ctx);
```
Comment (alanwaketan) [inline torch_xla/csrc/xla_graph_executor.cpp:1335]: Should we remove is_sharded in SetHloSharding?

Comment (alanwaketan) [reply torch_xla/csrc/xla_graph_executor.cpp:1335]: It looks like it's being used by other places. nvm.

**Review by alanwaketan — APPROVED** [review]
LGTM. Let's trigger the TPU CI to make sure there is no regression.

Not sure if we can add any test cases?

File: torch_xla/csrc/xla_graph_executor.cpp (hunk 1 above)
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:1335]: can we add a test for non-sharding execution with spmd mode?

Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.cpp:1335]: It looks like we already have some coverage here: https://github.com/pytorch/xla/blob/4f378e694a68192fd634a070a3485c73afab1c71/test/spmd/test_xla_virtual_device.py#L55

I'll sync up with @alanwaketan on the issue he encountered and use that as a test case

Comment (jonb377) [reply torch_xla/csrc/xla_graph_executor.cpp:1335]: Added a test case which induces the error without this patch

Comment (alanwaketan) [reply torch_xla/csrc/xla_graph_executor.cpp:1335]: Sorry for missing your messages, but the `Unknown device: SPMD:0` is the right error.

**Review by JackCaoG — APPROVED** [review]

**Comment by jonb377** [comment]
TPU CI passed, submitting this change.

**Pull request merged**